File: frontend/src/api/paper.ts
export type Paper = {
  id: number
  title: string
  authors: string
  journal: string
  year: number | null
  doi: string | null
  abstract: string | null
  keywords: string
  pdf_path: string | null
  created_at: string
  updated_at: string
}

export type PaperPayload = {
  title: string
  authors: string
  journal: string
  year: number | null
  doi: string | null
  abstract: string | null
  keywords: string
  pdf_path: string | null
}

const PAPERS_ENDPOINT = "/papers"

async function request<T>(url: string, options?: RequestInit): Promise<T> {
  const response = await fetch(url, {
    headers: {
      "Content-Type": "application/json",
      ...options?.headers,
    },
    ...options,
  })

  if (!response.ok) {
    const message = await response.text()
    throw new Error(message || `请求失败，状态码 ${response.status}`)
  }

  if (response.status === 204) {
    return undefined as T
  }

  return response.json() as Promise<T>
}

export function listPapers(): Promise<Paper[]> {
  return request<Paper[]>(PAPERS_ENDPOINT)
}

export function getPaper(id: number): Promise<Paper> {
  return request<Paper>(`${PAPERS_ENDPOINT}/${id}`)
}

export function searchPapers(keyword: string): Promise<Paper[]> {
  const params = new URLSearchParams({ keyword })
  return request<Paper[]>(`${PAPERS_ENDPOINT}/search?${params.toString()}`)
}

export function createPaper(payload: PaperPayload): Promise<Paper> {
  return request<Paper>(PAPERS_ENDPOINT, {
    method: "POST",
    body: JSON.stringify(payload),
  })
}

export function updatePaper(id: number, payload: PaperPayload): Promise<Paper> {
  return request<Paper>(`${PAPERS_ENDPOINT}/${id}`, {
    method: "PUT",
    body: JSON.stringify(payload),
  })
}

export async function uploadPaperPdf(id: number, file: File): Promise<Paper> {
  const formData = new FormData()
  formData.append("file", file)

  const response = await fetch(`${PAPERS_ENDPOINT}/${id}/upload`, {
    method: "POST",
    body: formData,
  })

  if (!response.ok) {
    const message = await response.text()
    throw new Error(message || `请求失败，状态码 ${response.status}`)
  }

  return response.json() as Promise<Paper>
}

export async function importPaperPdf(file: File): Promise<Paper> {
  const formData = new FormData()
  formData.append("file", file)

  const response = await fetch(`${PAPERS_ENDPOINT}/import`, {
    method: "POST",
    body: formData,
  })

  if (!response.ok) {
    const message = await response.text()
    throw new Error(message || `请求失败，状态码 ${response.status}`)
  }

  return response.json() as Promise<Paper>
}

export function deletePaper(id: number): Promise<void> {
  return request<void>(`${PAPERS_ENDPOINT}/${id}`, {
    method: "DELETE",
  })
}

File: backend/app/models/paper.py
from datetime import datetime, timezone
from typing import Optional

from sqlalchemy import DateTime, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base


def utc_now() -> datetime:
    return datetime.now(timezone.utc)


class Paper(Base):
    __tablename__ = "papers"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    title: Mapped[str] = mapped_column(String(255), nullable=False, index=True)
    authors: Mapped[str] = mapped_column(Text, nullable=False, default="")
    journal: Mapped[str] = mapped_column(String(255), nullable=False, default="")
    year: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    doi: Mapped[Optional[str]] = mapped_column(String(255), nullable=True, unique=True)
    abstract: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    keywords: Mapped[str] = mapped_column(Text, nullable=False, default="")
    pdf_path: Mapped[Optional[str]] = mapped_column(String(500), nullable=True)
    content: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=utc_now,
        nullable=False,
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=utc_now,
        onupdate=utc_now,
        nullable=False,
    )
